### PR TITLE
Add supported inference and incremental training configs

### PR DESCRIPTION
**Pull request by Captainia**
*Issue #, if available:*

*Description of changes:*

This change introduces supported inference config and supported incremental training config, and choose them correspondingly from the training config being used.

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/jumpstart/estimator.py [hunk 1]
```diff
@@ -112,6 +112,7 @@ def __init__(
         disable_output_compression: Optional[bool] = None,
         enable_remote_debug: Optional[Union[bool, PipelineVariable]] = None,
         config_name: Optional[str] = None,
+        inference_config_name: Optional[str] = None,
```
Comment (evakravi) [inline src/sagemaker/jumpstart/estimator.py:115]: this should not be exposed as a constructor argument imo

File: src/sagemaker/jumpstart/estimator.py [hunk 2]
```diff
@@ -504,7 +505,10 @@ def __init__(
             enable_remote_debug (bool or PipelineVariable): Optional.
                 Specifies whether RemoteDebug is enabled for the training job
             config_name (Optional[str]):
-                Name of the JumpStart Model config to apply. (Default: None).
+                Name of the training configuration to apply to the Estimator. (Default: None).
+            inference_config_name (Optional[str]):
```
Comment (evakravi) [inline src/sagemaker/jumpstart/estimator.py:509]: can we just have `inference_config_name` in `Estimator.deploy()` method?

File: src/sagemaker/jumpstart/factory/estimator.py [hunk 3]
```diff
@@ -793,3 +797,73 @@ def _add_fit_extra_kwargs(kwargs: JumpStartEstimatorFitKwargs) -> JumpStartEstim
...
+
+        resolved_config = specs.training_configs.configs[
+            kwargs.training_config_name
+        ].resolved_config
+        supported_instance_types = resolved_config.get("supported_training_instance_types", [])
+        if kwargs.instance_type not in supported_instance_types:
+            raise ValueError(
+                f"Instance type {kwargs.instance_type} "
+                f"is not supported for config {kwargs.training_config_name}."
+            )
```
Comment (evakravi) [inline src/sagemaker/jumpstart/factory/estimator.py:867]: this logic is hard to follow. can we split into smaller functions?

File: src/sagemaker/jumpstart/factory/estimator.py [hunk 4]
```diff
@@ -793,3 +797,73 @@ def _add_fit_extra_kwargs(kwargs: JumpStartEstimatorFitKwargs) -> JumpStartEstim
...
+            or specs.training_configs.get_top_config_from_ranking().default_incremental_trainig_config  # noqa E501  # pylint: disable=c0301
+        )
+
+    if specs.training_configs and specs.training_configs.get_top_config_from_ranking().config_name:
+        kwargs.training_config_name = (
+            kwargs.training_config_name
+            or specs.training_configs.get_top_config_from_ranking().config_name
+        )
+
+        kwargs.inference_config_name = (
```
Comment (evakravi) [inline src/sagemaker/jumpstart/factory/estimator.py:837]: this logic should be in the `model` factory. Can't you call `_get_deploy_kwargs` from the `estimator` factory to get the inference config name?

File: tests/unit/sagemaker/jumpstart/test_types.py [hunk 5]
```diff
@@ -1193,8 +1193,15 @@ def test_training_configs_parsing():
...
         {
+            "default_training_instance_type": "ml.trn1.2xlarge",
             "supported_training_instance_types": ["ml.trn1.xlarge", "ml.trn1.2xlarge"],
             "training_artifact_key": "artifacts/meta-textgeneration-llama-2-7b/neuron-training/model/",
+            "training_ecr_specs": {
+                "framework": "huggingface",
+                "framework_version": "2.0.0",
+                "py_version": "py310",
+                "huggingface_transformers_version": "4.28.1",
+            },
```
Comment (evakravi) [inline tests/unit/sagemaker/jumpstart/test_types.py:1204]: why do we need this?

File: src/sagemaker/jumpstart/estimator.py (hunk 1 above)
Comment (Captainia) [reply src/sagemaker/jumpstart/estimator.py:115]: Agreed, there is better way to do this. I moved to only expose in deploy method

File: src/sagemaker/jumpstart/factory/estimator.py (hunk 3 above)
Comment (Captainia) [reply src/sagemaker/jumpstart/factory/estimator.py:867]: Yes, a lot of these are not needed now since we moved them to deploy and attach method

File: tests/unit/sagemaker/jumpstart/test_types.py (hunk 5 above)
Comment (Captainia) [reply tests/unit/sagemaker/jumpstart/test_types.py:1204]: This is to satisfy parsing the default image for `neuron-training` config, otherwise a config component named neuron-training still has a `py3` version in the base spec

File: src/sagemaker/jumpstart/factory/model.py [hunk 6]
```diff
@@ -592,6 +614,45 @@ def _add_config_name_to_kwargs(kwargs: JumpStartModelInitKwargs) -> JumpStartMod
...
+    )
+
+    if training_config_name:
+        kwargs.config_name = _select_inference_config_from_training_config(
+            specs=specs, training_config_name=training_config_name
+        )
+
+    if (
+        specs.inference_configs
+        and specs.inference_configs.get_top_config_from_ranking().config_name
```
Comment (evakravi) [inline src/sagemaker/jumpstart/factory/model.py:648]: nit: can we store `specs.inference_configs.get_top_config_from_ranking().config_name` as a local variable?

Comment (Captainia) [reply src/sagemaker/jumpstart/factory/model.py]: Yea updated, thanks

File: src/sagemaker/jumpstart/factory/model.py [hunk 7]
```diff
@@ -643,13 +640,9 @@ def _add_config_name_to_deploy_kwargs(
...
         )
 
-    if (
-        specs.inference_configs
-        and specs.inference_configs.get_top_config_from_ranking().config_name
-    ):
-        kwargs.config_name = (
-            kwargs.config_name or specs.inference_configs.get_top_config_from_ranking().config_name
-        )
+    if specs.inference_configs and specs.inference_configs.get_top_config_from_ranking():
```
Comment (evakravi) [inline src/sagemaker/jumpstart/factory/model.py:643]: it seems you just got rid of `.config_name` from the if-statement

**Review by evakravi — APPROVED** [review]

File: src/sagemaker/jumpstart/factory/model.py (hunk 7 above)
Comment (Captainia) [reply src/sagemaker/jumpstart/factory/model.py]: sorry extracted to `default_config_name` and removed the redundant check `..get_top_config_from_ranking()` in the if clause. This is not needed since it fallbacks to None anyway

**Review by zhaoqizqwang — APPROVED** [review]
